File: src/camera.rs
use crate::{ray::Ray, Point3, Vec3};
use fastrand::Rng;

pub struct Camera {
    origin: Point3,
    lower_left_corner: Point3,
    horizontal: Vec3,
    vertical: Vec3,
    u: Vec3,
    v: Vec3,
    lens_radius: f32,
}

impl Camera {
    pub fn new(
        lookfrom: Point3,
        lookat: Point3,
        vup: Vec3,
        vfov: f32,
        aspect_ratio: f32,
        aperture: f32,
        focus_dist: f32,
    ) -> Self {
        let theta = vfov.to_radians();
        let h = (theta * 0.5).tan();
        let viewport_height = 2.0 * h;
        let viewport_width = aspect_ratio * viewport_height;

        let w = (lookfrom - lookat).normalize();
        let u = vup.cross(w).normalize();
        let v = w.cross(u);

        let origin = lookfrom;
        let horizontal = u * viewport_width * focus_dist;
        let vertical = v * viewport_height * focus_dist;
        let lower_left_corner =
            origin - horizontal * 0.5 - vertical * 0.5 - w * focus_dist;

        let lens_radius = aperture * 0.5;

        Self {
            origin,
            lower_left_corner,
            horizontal,
            vertical,
            u,
            v,
            lens_radius,
        }
    }

    pub fn get_ray(&self, s: f32, t: f32, rng: &Rng) -> Ray {
        let rd = crate::random::in_unit_disk(rng) * self.lens_radius;
        let offset = self.u * rd.x + self.v * rd.y;

        Ray {
            origin: self.origin + offset,
            dir: self.lower_left_corner
                + self.horizontal * s
                + self.vertical * t
                - self.origin
                - offset,
        }
    }
}

File: lua/raytracer/camera.lua
Camera = {}

function Camera.new(obj)
    assert(obj.look_from ~= nil)
    assert(obj.look_at ~= nil)
    assert(obj.vup ~= nil)
    assert(obj.vfov ~= nil)
    assert(obj.aperture ~= nil)
    assert(obj.focus_dist ~= nil)

    setmetatable(obj, Camera)

    return obj
end

function Camera:__tostring()
    return string.format([[
Camera (
	look_from: %s,
	look_at: %s,
	vup: %s,
	vfov: %s,
	aperture: %s,
	focus_dist: %s,
)]], self.look_from, self.look_at, self.vup, self.vfov, self.aperture,
                         self.focus_dist)
end
